domain/domain.go

```diff
@@ -546,41 +548,69 @@ func (do *Domain) UpdateTableStatsLoop(ctx context.Context) error {
 	if lease <= 0 {
 		return nil
 	}
-	deltaUpdateDuration := lease * 5
-	go func(do *Domain) {
-		loadTicker := time.NewTicker(lease)
-		defer loadTicker.Stop()
-		deltaUpdateTicker := time.NewTicker(deltaUpdateDuration)
-		defer deltaUpdateTicker.Stop()
+	go do.updateStatsWorker(ctx, lease)
+	go do.autoAnalyzeWorker(lease)
+	return nil
+}
 
-		for {
-			select {
-			case <-loadTicker.C:
-				err = statsHandle.Update(do.InfoSchema())
-				if err != nil {
-					log.Error(errors.ErrorStack(err))
-				}
-			case <-do.exit:
-				return
+func (do *Domain) updateStatsWorker(ctx context.Context, lease time.Duration) {
+	deltaUpdateDuration := lease * 5
+	loadTicker := time.NewTicker(lease)
+	defer loadTicker.Stop()
+	deltaUpdateTicker := time.NewTicker(deltaUpdateDuration)
+	defer deltaUpdateTicker.Stop()
+	statsHandle := do.StatsHandle()
+	for {
+		select {
+		case <-loadTicker.C:
+			err := statsHandle.Update(do.InfoSchema())
+			if err != nil {
+				log.Error("[stats] update stats info fail: ", errors.ErrorStack(err))
+			}
+		case <-do.exit:
+			return
 			// This channel is sent only by ddl owner or the drop stats executor.
-			case t := <-statsHandle.DDLEventCh():
-				err = statsHandle.HandleDDLEvent(t)
+		case t := <-statsHandle.DDLEventCh():
+			err := statsHandle.HandleDDLEvent(t)
+			if err != nil {
+				log.Error("[stats] handle ddl event fail: ", errors.ErrorStack(err))
+			}
+		case t := <-statsHandle.AnalyzeResultCh():
+			for _, hg := range t.Hist {
+				err := hg.SaveToStorage(ctx, t.TableID, t.Count, t.IsIndex)
 				if err != nil {
-					log.Error(errors.ErrorStack(err))
-				}
-			case t := <-statsHandle.AnalyzeResultCh():
-				for _, hg := range t.Hist {
-					err = hg.SaveToStorage(t.Ctx, t.TableID, t.Count, t.IsIndex)
-					if err != nil {
-						log.Error(errors.ErrorStack(err))
-					}
+					log.Error("[stats] save histogram to storage fail: ", errors.ErrorStack(err))
 				}
-			case <-deltaUpdateTicker.C:
-				statsHandle.DumpStatsDeltaToKV()
 			}
+		case <-deltaUpdateTicker.C:
+			statsHandle.DumpStatsDeltaToKV()
 		}
-	}(do)
-	return nil
+	}
+}
+
+func (do *Domain) autoAnalyzeWorker(lease time.Duration) {
+	id := do.ddl.OwnerManager().ID()
+	cancelCtx, cancelFunc := goctx.WithCancel(goctx.Background())
+	var statsOwner owner.Manager
+	if do.etcdClient == nil {
+		statsOwner = owner.NewMockManager(id, cancelFunc)
+	} else {
+		statsOwner = owner.NewOwnerManager(do.etcdClient, statistics.StatsPrompt, id, statistics.StatsOwnerKey, cancelFunc)
+	}
+	err := statsOwner.CampaignOwner(cancelCtx)
```

Comment:
It‘s better to put it out of a goroutine. If it hard to handle, you can add a `TODO` here.